Upgrade to Salt 3000.2 python2

TeddyAndrieux opened the pull request:
**Component**:

'salt'
<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: 

Two CVE were found on the Salt side that has a huge security impact, Salt version 3000.2 include these fixes.

https://docs.saltstack.com/en/latest/topics/releases/3000.2.html#salt-3000-2-release-notes
https://labs.f-secure.com/advisories/saltstack-authorization-bypass

**Summary**:

In order to fix this CVE ASAP do not wait to have a full python3 solution for MetalK8s and use the python2 version of Salt 3000.2.

To support Salt 3000.2:
- Handle "non-string" version in our custom package manager Salt module
- Apply `tojson` filter when dumping certs from Salt mine to avoid unicode invalid certificate
- Changes mine send and mine functions in pillar to do not use kwargs as it does not work in Salt 3000.2 (https://github.com/saltstack/salt/issues/56584)
- backport https://github.com/scality/metalk8s/commit/0dbcbd18b5668d5a833107f43c6e8e8105c91df8 in 2.0
- Remove `token_type` from `kubernetes_rbac` salt Auth as we cannot provide additional kwargs for auth in Salt 3000.2 (https://github.com/saltstack/salt/commit/3dbe8dc8bed169ca35edee64c421f585f6263dad)

---

Fixes: #650

NicolasT committed:
salt: use `tojson` when splicing in complex structures

Sometimes we use a complex structure as an interpolated value in an SLS
file. Whilst this works, this is merely by accident, and Salt 2019.2
changes the behaviour: complex values would be spliced in using their
`unicode` representation, including `u""` markers.

As such, use a proper `tojson` filter in the Jinja templates instead.

See: #650
See: https://github.com/scality/metalk8s/issues/650
See: https://docs.saltstack.com/en/latest/topics/releases/2019.2.0.html#non-backward-compatible-change-to-yaml-renderer
(cherry picked from commit 0dbcbd18b5668d5a833107f43c6e8e8105c91df8)

TeddyAndrieux committed:
salt: `metalk8s_packages_manager` module handle non-string version

Package version may not be a string but an int or a float, so just cast
the versions to string before comparing them

TeddyAndrieux committed:
salt: Apply `tojson` filter on CA advertised

In salt we push CA content to the mine but in newer Salt version
(>= 2019) if we not use the `tojson` filter on it we get unicode lines
so on each line of the resulting certificate you will have
`u'<content>'` which is not a valid certificate

TeddyAndrieux committed:
build: Upgrade Salt to `3000.2`

Change salt-master image to salt 3000.2 and download salt-minion version
3000.2

Fixes: #650

TeddyAndrieux committed:
salt: Fix mine call for salt '3000.2'

In Salt '3000.2' we can't provide kwargs for mine call, so we need to
use potitional arguments in pillar and for `module.run` with `mine.send`
we need to use potitional arguments to provide the mine name, as `func`
get renamed to `name` but this arguments is gave to the function behind,
which make the mine function call fail.

Sees https://github.com/saltstack/salt/issues/56584

bert-e commented:
# Hello teddyandrieux,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Branches have diverged



This pull request's source branch `improvement/GH-650-upgrade-to-salt-3000` has diverged from
`development/2.6` by more than 50 commits.

To avoid any integration risks, please re-synchronize them using one of the
following solutions:

* Merge `origin/development/2.6` into `improvement/GH-650-upgrade-to-salt-3000`
* Rebase `improvement/GH-650-upgrade-to-salt-3000` onto `origin/development/2.6`

Note: If you choose to rebase, you may have to ask me to rebuild
integration branches using the `reset` command.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/GH-650-upgrade-to-salt-3000` with contents from `improvement/GH-650-upgrade-to-salt-3000`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/GH-650-upgrade-to-salt-3000 origin/development/2.1
 $ git merge origin/improvement/GH-650-upgrade-to-salt-3000
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/GH-650-upgrade-to-salt-3000
```

TeddyAndrieux committed:
salt: Remove `token_type` from `kubernetes_rbac` salt Auth

Salt auth not longer support additional kwargs for custom auth in Salt
3000.2, so we cannot provide `token_type` instead consider it's a Bearer
auth if a `token` is provided and a Basic auth if we have a `password`

See: https://github.com/saltstack/salt/commit/3dbe8dc8bed169ca35edee64c421f585f6263dad

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/improvement/GH-650-upgrade-to-salt-3000` with contents from `w/2.3/improvement/GH-650-upgrade-to-salt-3000`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/improvement/GH-650-upgrade-to-salt-3000 origin/development/2.4
 $ git merge origin/w/2.3/improvement/GH-650-upgrade-to-salt-3000
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/improvement/GH-650-upgrade-to-salt-3000
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/GH-650-upgrade-to-salt-3000` with contents from `w/2.4/improvement/GH-650-upgrade-to-salt-3000`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/GH-650-upgrade-to-salt-3000 origin/development/2.5
 $ git merge origin/w/2.4/improvement/GH-650-upgrade-to-salt-3000
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/GH-650-upgrade-to-salt-3000
```

bert-e commented:
# History mismatch



Merge commit #19ef077932b64e9be95ba7496b7905cba27b982a on the integration branch
`w/2.1/improvement/GH-650-upgrade-to-salt-3000` is merging a branch which is neither the current
branch `improvement/GH-650-upgrade-to-salt-3000` nor the development branch
`development/2.1`.

It is likely due to a rebase of the branch `improvement/GH-650-upgrade-to-salt-3000` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

TeddyAndrieux commented:
/reset

bert-e commented:
# Reset complete



I have successfully deleted this pull request's integration branches.

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.1/improvement/GH-650-upgrade-to-salt-3000` with contents from `improvement/GH-650-upgrade-to-salt-3000`
and `development/2.1`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.1/improvement/GH-650-upgrade-to-salt-3000 origin/development/2.1
 $ git merge origin/improvement/GH-650-upgrade-to-salt-3000
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.1/improvement/GH-650-upgrade-to-salt-3000
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.4/improvement/GH-650-upgrade-to-salt-3000` with contents from `w/2.3/improvement/GH-650-upgrade-to-salt-3000`
and `development/2.4`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.4/improvement/GH-650-upgrade-to-salt-3000 origin/development/2.4
 $ git merge origin/w/2.3/improvement/GH-650-upgrade-to-salt-3000
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.4/improvement/GH-650-upgrade-to-salt-3000
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.5/improvement/GH-650-upgrade-to-salt-3000` with contents from `w/2.4/improvement/GH-650-upgrade-to-salt-3000`
and `development/2.5`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.5/improvement/GH-650-upgrade-to-salt-3000 origin/development/2.5
 $ git merge origin/w/2.4/improvement/GH-650-upgrade-to-salt-3000
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.5/improvement/GH-650-upgrade-to-salt-3000
```

bert-e commented:
# Conflict




A conflict has been raised during the creation of
integration branch `w/2.6/improvement/GH-650-upgrade-to-salt-3000` with contents from `w/2.5/improvement/GH-650-upgrade-to-salt-3000`
and `development/2.6`.

**I have not created the integration branch.**


Here are the steps to resolve this conflict:

```bash
 $ git fetch
 $ git checkout -B w/2.6/improvement/GH-650-upgrade-to-salt-3000 origin/development/2.6
 $ git merge origin/w/2.5/improvement/GH-650-upgrade-to-salt-3000
 $ # <intense conflict resolution>
 $ git commit
 $ git push -u origin w/2.6/improvement/GH-650-upgrade-to-salt-3000
```

bert-e commented:
# History mismatch



Merge commit #604a2922455cebaaa856ccf2158521dda075da45 on the integration branch
`w/2.5/improvement/GH-650-upgrade-to-salt-3000` is merging a branch which is neither the current
branch `improvement/GH-650-upgrade-to-salt-3000` nor the development branch
`development/2.5`.

It is likely due to a rebase of the branch `improvement/GH-650-upgrade-to-salt-3000` and the
merge is not possible until all related `w/*` branches are deleted or updated.

**Please use the `reset` command to have me reinitialize these branches.**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

TeddyAndrieux commented:
/help

bert-e commented:
# Help page



The following options and commands are available at this time.

## Options ##


name   | description  | privileged   | authored
------ | ------------ | ------------ |----------
:arrow_right: **bypass_leader_approval** | Bypass the pull request leaders' approval |  :star:  | 
:arrow_right: **create_pull_requests** | Allow the creation of integration pull requests. |  | 
:arrow_right: **unanimity** | Change review acceptance criteria from `one reviewer at least` to `all reviewers` |  | 
:arrow_right: **approve** | Instruct Bert-E that the author has approved the pull request. |  |  :writing_hand: 
:arrow_right: **bypass_jira_check** | Bypass the Jira issue check |  :star:  | 
:arrow_right: **wait** | Instruct Bert-E not to run until further notice. |  | 
:arrow_right: **bypass_peer_approval** | Bypass the pull request peers' approval |  :star:  | 
:arrow_right: **bypass_commit_size** | Bypass the check on the size of the changeset ```TBA``` |  :star:  | 
:arrow_right: **bypass_build_status** | Bypass the build and test status |  :star:  | 
:arrow_right: **after_pull_request** | Wait for the given pull request id to be merged before continuing with the current one. |  | 
:arrow_right: **bypass_incompatible_branch** | Bypass the check on the source branch prefix |  :star:  | 
:arrow_right: **no_octopus** | Prevent Wall-E from doing any octopus merge and use multiple consecutive merge instead |  | 
:arrow_right: **bypass_author_approval** | Bypass the pull request author's approval |  :star:  | 



## Commands ##


name   | description  | privileged
------ | ------------ | ------------
:arrow_right: **force_reset** | Delete integration branches & pull requests, and restart merge process from the beginning. | 
:arrow_right: **build** | Re-start a fresh build ```TBA``` | 
:arrow_right: **retry** | Re-start a fresh build ```TBA``` | 
:arrow_right: **clear** | Remove all comments from Bert-E from the history ```TBA``` | 
:arrow_right: **help** | Print Bert-E's manual in the pull request. | 
:arrow_right: **reset** | Try to remove integration branches unless there are commits on them which do not appear on the source branch. | 
:arrow_right: **status** | Print Bert-E's current status in the pull request ```TBA``` |

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote

TeddyAndrieux commented:
/create_pull_requests

bert-e commented:
# Integration data created



I have created the integration data for the additional destination branches.

* this pull request will merge `improvement/GH-650-upgrade-to-salt-3000` into
`development/2.0`
* [w/2.1/improvement/GH-650-upgrade-to-salt-3000](https://github.com/scality/metalk8s/compare/development/2.1...w/2.1/improvement/GH-650-upgrade-to-salt-3000) will be merged into `development/2.1` (pull request #2514)
* [w/2.2/improvement/GH-650-upgrade-to-salt-3000](https://github.com/scality/metalk8s/compare/development/2.2...w/2.2/improvement/GH-650-upgrade-to-salt-3000) will be merged into `development/2.2` (pull request #2515)
* [w/2.3/improvement/GH-650-upgrade-to-salt-3000](https://github.com/scality/metalk8s/compare/development/2.3...w/2.3/improvement/GH-650-upgrade-to-salt-3000) will be merged into `development/2.3` (pull request #2516)
* [w/2.4/improvement/GH-650-upgrade-to-salt-3000](https://github.com/scality/metalk8s/compare/development/2.4...w/2.4/improvement/GH-650-upgrade-to-salt-3000) will be merged into `development/2.4` (pull request #2517)
* [w/2.5/improvement/GH-650-upgrade-to-salt-3000](https://github.com/scality/metalk8s/compare/development/2.5...w/2.5/improvement/GH-650-upgrade-to-salt-3000) will be merged into `development/2.5` (pull request #2518)
* [w/2.6/improvement/GH-650-upgrade-to-salt-3000](https://github.com/scality/metalk8s/compare/development/2.6...w/2.6/improvement/GH-650-upgrade-to-salt-3000) will be merged into `development/2.6` (pull request #2519)



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`




*Follow* integration pull requests if you would like to be notified of
build statuses by email.







*The following options are set:* **create_pull_requests**

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @nootal

* @alexandre-allard-scality

* @lucieleonard

* @sayf-eddine-scality

* @slaperche-scality

* @nicolast

* @gdemonet

* @carlito-scality

* @thomasdanan

* @jbertran

* @teddyandrieux

* @cmonfort

* @monpote










*The following options are set:* **create_pull_requests**

gdemonet commented on salt/_auth/kubernetes_rbac.py:
```diff
@@ -147,20 +147,39 @@ def _load_kubeconfig(opts):
+            '"password" at the same time'
+        )
+        return False
+    if not token and (not password or not username):
```
Don't change it, but in the future, you can make this boolean expression more readable with:

```python
if not (token or password and username):

# Or with parentheses
if not (token or (password and username)):
```

NicolasT replied:
An underrated pattern:
```python
if not any([
        token,
        username and password,
        ]):
    blah()
```

gdemonet replied:
Like this one, it's even more readable :)

NicolasT replied:
While strictly speaking less efficient (so could make a difference in 'hot' loops and such), `any` and `all` often make multiple `or`'ed or `and`'ed boolean expressions easier to grasp.

gdemonet reviewed (approved):
LGTM! Let's fix the upgrade and downgrade issues in a follow-up PR (can you open tickets so we can track them please?)

TeddyAndrieux commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **create_pull_requests, approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.0`

* :heavy_check_mark: `development/2.1`

* :heavy_check_mark: `development/2.2`

* :heavy_check_mark: `development/2.3`

* :heavy_check_mark: `development/2.4`

* :heavy_check_mark: `development/2.5`

* :heavy_check_mark: `development/2.6`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`



Please check the status of the associated issue GH-650.

Goodbye teddyandrieux.
